MHVLibTests/Clients/MHVThingClientBlobTests.m

```diff
@@ -0,0 +1,502 @@
+                
+                let(blobDownloadRequest, ^{
+                    return (MHVBlobDownloadRequest *)requestedServiceOperationStep1;
+                });
+                
+                it(@"should use correct URL", ^
+                   {
+                       [[blobDownloadRequest.url.absoluteString should] equal:@"http://blob.test/path/blob"];
+                   });
+                
+                it(@"should have file path", ^
+                   {
+                       [[blobDownloadRequest.toFilePath should] equal:@"//to/path/name.xyz"];
+                   });
+                
+                it(@"should be an anonymous request", ^
+                   {
+                       [[theValue(blobDownloadRequest.isAnonymous) should] beYes];
+                   });
+            });
+    
+    context(@"DownloadBlob Data Inline", ^
+            {
+                __block NSData *returnedData;
+                beforeAll(^{
+                    requestedServiceOperationStep1 = nil;
+                    requestedServiceOperationStep2 = nil;
+                    
+                    MHVBlobPayloadThing *blobPayload = [[MHVBlobPayloadThing alloc] init];
+                    blobPayload.inlineData = [@"123456" dataUsingEncoding:NSUTF8StringEncoding];
+                    
+                    [thingClient downloadBlobData:blobPayload
+                                       completion:^(NSData * _Nullable data, NSError * _Nullable error)
+                     {
+                         returnedData = data;
+                     }];
+                });
+                
+                it(@"should return data", ^
+                   {
+                       [[expectFutureValue(returnedData) shouldEventually] beNonNil];
+                       
+                       NSString *dataString = [[NSString alloc] initWithData:returnedData encoding:NSUTF8StringEncoding];
+                       
+                       [[dataString should] equal:@"123456"];
+                   });
+            });
+    
+    context(@"DownloadBlob Errors", ^
+            {
+                beforeAll(^{
+                    requestedServiceOperationStep1 = nil;
+                    requestedServiceOperationStep2 = nil;
+                });
+                
+                it(@"should fail if blob payload is nil", ^
+                   {
+                       __block NSError *requestError;
+                       [thingClient downloadBlobData:nil
+                                          completion:^(NSData * _Nullable data, NSError * _Nullable error)
+                        {
+                            requestError = error;
+                        }];
+                       
+                       [[requestError should] beNonNil];
+                       [[theValue(requestError.code) should] equal:@(MHVErrorTypeRequiredParameter)];
+                   });
+                
+            });
+
+    context(@"UploadBlob step 1", ^
```

Comment:
We probably shouldn't have tests that specifically call out the implementation details of the service. We should test the final result of the call which is either success or error. We can assume that all requirements are met and all of the method calls are successful, the result to the caller is success and the output from the method can be validated. However, if errors occur along the way web should verify the output contains a useful error (an error that specifically calls out what went wrong) and that the 'thing' parameter is nil. 'Step 1' is only step one because of the current implementation - If this changes in the future these tests are broken.